### PR TITLE
Add shortcutting for boolean operations

### LINKED ISSUE #387: Make && and || by name
We can make these two logical operators by name, by simply desugaring them to `if`.

This could either happen before Typer (that is in Parser), or after Typer (that is in the translation to core).

```
EXPR1 || EXPR2 = if (EXPR1) true else EXPR2
EXPR1 && EXPR2 = if (EXPR1) EXPR2 else false
```

Doing it in Parser has the advantage that it is really just a few lines of code to implement.
Doing it in core.Transformer has the advantage that we can selectively perform this translation on `effekt.infixOr` and `effekt.InfixAnd` and the operators can be overloaded (for instance for logic programming).

We could also do the translation selectively and **only** translate it if EXPR2 is a Stmt or a side effecting Expr (EXPR1 does not matter, since it is forced anyway).

### PR DESCRIPTION
**Pull request by marvinborner**
This PR closes #387. For now I've implemented it directly in the parser.
There might be bugs left, so please try it out if the CI passes.

---

CI did not pass. LLVM has obvious errors that could be fixed by improving the inliner.

### CODE REVIEW

**Comment by b-studios** [comment]
It seems the inliner is turned off?

https://github.com/effekt-lang/effekt/blob/62a551e0c8b5090a8e376a3549abde5f9781e40d/effekt/shared/src/main/scala/effekt/core/Optimizer.scala#L24

**Comment by b-studios** [comment]
Looks like @dvdvgt disabled it here: fc29e654

**Comment by dvdvgt** [comment]
This was disabled for debugging during the development of the accompanying PR. This can, of course, be turned back on!

File: effekt/shared/src/main/scala/effekt/Parser.scala [hunk 1]
```diff
@@ -617,9 +617,15 @@ class EffektParsers(positions: Positions) extends EffektLexers(positions) {
   private def binaryOp(lhs: Term, op: String, rhs: Term): Term =
      Call(IdTarget(IdRef(Nil, opName(op))), Nil, List(lhs, rhs), Nil)
 
-  private def opName(op: String): String = op match {
+  private def liftedBinaryOp(lhs: Term, op: String, rhs: Term): Term =
```
Comment (b-studios) [inline effekt/shared/src/main/scala/effekt/Parser.scala:620]: I would call this `delayedBinaryOp` or `thunkedBinaryOp`. We already use "lifted" for other things a lot :)

Comment (b-studios) [reply effekt/shared/src/main/scala/effekt/Parser.scala:620]: Also, could you give an example in pseudocode like

```
delayedBinaryOp(lhs, op, rhs) = op { lhs } { rhs }
```
?

File: libraries/chez/callcc/effekt.effekt [hunk 2]
```diff
@@ -141,11 +141,11 @@ extern pure def infixGte(x: Double, y: Double): Bool =
 extern pure def not(b: Bool): Bool =
   "(not ${b})"
 
-extern pure def infixOr(x: Bool, y: Bool): Bool =
-  "(or ${x} ${y})"
+def infixOr { first:  () => Bool } { second: () => Bool }: Bool =
```
Comment (b-studios) [inline libraries/chez/callcc/effekt.effekt:144]: You could use the shorthand here

```suggestion
def infixOr { first:  => Bool } { second: => Bool }: Bool = 
```

which emphasizes the thunkiness

**Comment by b-studios** [comment]
Just out of curiosity. Could you post the generated code (js, chez, or llvm) of a simple example (after rebasing on the master with the inliner fix)?

It would be interesting to see whether the "overhead" of the function call and all traces of the "monadic" encoding are completely removed.

**Comment by marvinborner** [comment]
I can't reproduce the [CI failure](https://github.com/effekt-lang/effekt/actions/runs/8571765961/job/23492781723) on `stdlib/list/insert.effekt` of `chez-callcc`. The [previous CI run](https://github.com/effekt-lang/effekt/actions/runs/8521298424/job/23339157347) passed (although with a slightly different, yet equivalent, git history).
Could this be related to our other flakiness issues? Otherwise I wouldn't know which of my changes causes this.

**Review by b-studios — APPROVED** [review]
Beautiful, thanks!

**Pull request merged**